modules/tftp/http_downloads.rb

```diff
@@ -0,0 +1,318 @@
+    end
+
+    def status(id)
+      m.synchronize { id_to_download.key?(id) ? id_to_download[id].status : nil }
+    end
+
+    def restart_downloads
+      interrupted_downloads = Dir.glob(File.join(@dst_dir, '**', "*.metadata"))
+      interrupted_downloads.each do |path|
+        begin
+          filepath = remove_file_extention(remove_file_extention(path))
+          tmp_filepath = remove_file_extention(path) + '.tmp'
+          parsed_metadata = JSON.parse(IO.read(path))
+          do_download(filepath, tmp_filepath, path, parsed_metadata['url'], parsed_metadata['id'], true)
+        rescue Exception => e
+          logger.error("Error restarting download '#{filepath}'", e)
+        end
+      end
+    end
+
+    def download(prefix, src_url)
+      filepath = filepath(prefix, src_url)
+      raise "TFTP destination outside of tftproot" unless filepath.to_s.start_with?(@dst_dir)
+      tmp_filepath = filepath + tmp_filepath_postfix(src_url)
+      metadata_filepath = filepath + metadata_filepath_postfix(src_url)
+      return unless downloader_class.new(tmp_filepath, src_url).is_local_copy_stale?(filepath)
+      do_download(filepath, tmp_filepath, metadata_filepath, src_url)
+    end
+
+    def do_download(filepath, tmp_filepath, metadata_filepath, src_url, id = SecureRandom.hex(12), restart = false)
+      m.synchronize do
+        begin
+          return url_to_download[src_url] if url_to_download.key?(src_url)
+          unless restart
+            FileUtils.mkdir_p(Pathname.new(filepath).parent)
+            create_metadata(metadata_filepath, src_url, id)
+          end
+          start_or_restart_download(filepath, tmp_filepath, metadata_filepath, src_url, id, restart)
+          id
+        rescue Exception => e
+          logger.debug("Error creating metadata for download from '#{src_url}'", e)
+          remove_file(metadata_filepath)
+          remove_file(tmp_filepath)
+          raise e
+        end
+      end
+    end
+
+    def start_or_restart_download(filepath, tmp_filepath, metadata_filepath, src_url, id, restart)
+      download_in_progress = downloader_class.new(tmp_filepath, src_url)
+
+      url_to_download[src_url] = id
+      id_to_download[id] = download_in_progress
+
+      @download_in_progress =
+        (restart ? download_in_progress.restart_download : download_in_progress.download)
+        .on_success do
+          unless download_in_progress.status.stopped?
+            logger.debug("Successfully downloaded '#{src_url}'")
+            File.rename(tmp_filepath, filepath)
+            remove_file(metadata_filepath)
+          end
+        end
+        .on_error do |reason|
+          logger.debug("Error when downloading '#{src_url}'", reason)
+          download_in_progress.status.last_error = reason
+          remove_file(tmp_filepath)
+          remove_file(metadata_filepath)
+        end
+        .then { m.synchronize { url_to_download.delete(src_url) }}
```

Comment:
> Will the then block be executed if error occurs?

yes.

> Also, url_to_download is synchronized here, but not all manipulation is synchronized (e.g. line 254 above).

```start_or_restart_download``` is only called from under a mutex, none of the calls here need synchronization (or can use it, as it is a mutex). The promise created in ```#then``` is executed in a different thread, and needs a dedicated synchronization.

This does warrant a comment, however.